File: src/weekly_intel/ingestion/youtube.py
"""YouTube transcript ingestor."""

import hashlib
import logging
import re
from datetime import datetime
from typing import Optional
from urllib.parse import parse_qs, urlparse

import httpx
from youtube_transcript_api import YouTubeTranscriptApi
from youtube_transcript_api._errors import NoTranscriptFound, TranscriptsDisabled

from weekly_intel.database import ContentItem, get_session
from weekly_intel.ingestion.base import BaseIngestor, ContentData, IngestResult

logger = logging.getLogger(__name__)


class YouTubeIngestor(BaseIngestor):
    """Ingestor for YouTube video transcripts."""

    source_type = "youtube"

    def validate_config(self) -> list[str]:
        """Validate YouTube configuration."""
        errors = []

        has_channel = self.config.get("channel_id") or self.config.get("channel_url")
        has_playlist = self.config.get("playlist_id") or self.config.get("playlist_url")
        has_videos = self.config.get("video_urls") or self.config.get("video_ids")

        if not (has_channel or has_playlist or has_videos):
            errors.append(
                "YouTube source requires one of: channel_id, channel_url, "
                "playlist_id, playlist_url, video_urls, or video_ids"
            )

        return errors

    def _extract_video_id(self, url: str) -> Optional[str]:
        """Extract video ID from various YouTube URL formats."""
        # Handle different URL formats
        patterns = [
            r"(?:v=|/v/|youtu\.be/)([a-zA-Z0-9_-]{11})",
            r"(?:embed/)([a-zA-Z0-9_-]{11})",
            r"(?:shorts/)([a-zA-Z0-9_-]{11})",
        ]

        for pattern in patterns:
            match = re.search(pattern, url)
            if match:
                return match.group(1)

        return None

    def _extract_channel_id(self, url: str) -> Optional[str]:
        """Extract channel ID from URL."""
        # Handle /channel/CHANNEL_ID format
        match = re.search(r"/channel/([a-zA-Z0-9_-]+)", url)
        if match:
            return match.group(1)

        # Handle /@username format - would need API call to resolve
        # For now, return None and let the user provide channel_id directly
        return None

    def _extract_playlist_id(self, url: str) -> Optional[str]:
        """Extract playlist ID from URL."""
        parsed = urlparse(url)
        params = parse_qs(parsed.query)
        return params.get("list", [None])[0]

    def _get_video_ids(self) -> list[str]:
        """Get list of video IDs to process."""
        video_ids = []

        # Direct video IDs
        if self.config.get("video_ids"):
            video_ids.extend(self.config["video_ids"])

        # Video URLs
        if self.config.get("video_urls"):
            for url in self.config["video_urls"]:
                vid = self._extract_video_id(url)
                if vid:
                    video_ids.append(vid)

        # Playlist (requires scraping or API)
        playlist_id = self.config.get("playlist_id")
        if not playlist_id and self.config.get("playlist_url"):
            playlist_id = self._extract_playlist_id(self.config["playlist_url"])

        if playlist_id:
            playlist_videos = self._fetch_playlist_videos(playlist_id)
            video_ids.extend(playlist_videos)

        # Channel (requires scraping or API)
        channel_id = self.config.get("channel_id")
        if not channel_id and self.config.get("channel_url"):
            channel_id = self._extract_channel_id(self.config["channel_url"])

        if channel_id:
            channel_videos = self._fetch_channel_videos(channel_id)
            video_ids.extend(channel_videos)

        # Deduplicate while preserving order
        seen = set()
        unique_ids = []
        for vid in video_ids:
            if vid not in seen:
                seen.add(vid)
                unique_ids.append(vid)

        # Apply limit
        max_videos = self.config.get("max_videos", 20)
        return unique_ids[:max_videos]

    def _fetch_playlist_videos(self, playlist_id: str) -> list[str]:
        """Fetch video IDs from a playlist using web scraping."""
        logger.info(f"Fetching playlist: {playlist_id}")

        # Use YouTube's oembed/API-less approach
        # This is a simplified version - in production you'd use the Data API
        url = f"https://www.youtube.com/playlist?list={playlist_id}"

        try:
            with httpx.Client(timeout=30) as client:
                response = client.get(url)
                response.raise_for_status()

                # Extract video IDs from the page
                video_ids = re.findall(r'"videoId":"([a-zA-Z0-9_-]{11})"', response.text)
                unique_ids = list(dict.fromkeys(video_ids))  # Deduplicate preserving order

                logger.info(f"Found {len(unique_ids)} videos in playlist")
                return unique_ids[:50]  # Limit

        except Exception as e:
            logger.error(f"Failed to fetch playlist: {e}")
            return []

    def _fetch_channel_videos(self, channel_id: str) -> list[str]:
        """Fetch recent video IDs from a channel."""
        logger.info(f"Fetching channel: {channel_id}")

        # Use RSS feed for channel videos (doesn't require API key)
        rss_url = f"https://www.youtube.com/feeds/videos.xml?channel_id={channel_id}"

        try:
            import feedparser

            feed = feedparser.parse(rss_url)
            video_ids = []

            for entry in feed.entries:
                # Extract video ID from entry link
                vid = self._extract_video_id(entry.get("link", ""))
                if vid:
                    video_ids.append(vid)

            logger.info(f"Found {len(video_ids)} videos from channel RSS")
            return video_ids

        except Exception as e:
            logger.error(f"Failed to fetch channel: {e}")
            return []

    def _fetch_transcript(self, video_id: str) -> Optional[str]:
        """Fetch transcript for a video."""
        try:
            transcript_list = YouTubeTranscriptApi.list_transcripts(video_id)

            # Prefer manual transcripts, fallback to auto-generated
            transcript = None
            try:
                transcript = transcript_list.find_manually_created_transcript(["en"])
            except NoTranscriptFound:
                try:
                    transcript = transcript_list.find_generated_transcript(["en"])
                except NoTranscriptFound:
                    # Try any available transcript
                    for t in transcript_list:
                        transcript = t
                        break

            if transcript:
                transcript_data = transcript.fetch()
                # Combine all text segments
                text = " ".join(segment["text"] for segment in transcript_data)
                return text

        except TranscriptsDisabled:
            logger.warning(f"Transcripts disabled for video {video_id}")
        except Exception as e:
            logger.warning(f"Failed to get transcript for {video_id}: {e}")

        return None

    def _fetch_video_metadata(self, video_id: str) -> dict:
        """Fetch video metadata using oembed."""
        try:
            url = f"https://www.youtube.com/oembed?url=https://www.youtube.com/watch?v={video_id}&format=json"
            with httpx.Client(timeout=10) as client:
                response = client.get(url)
                if response.status_code == 200:
                    return response.json()
        except Exception as e:
            logger.warning(f"Failed to get metadata for {video_id}: {e}")

        return {}

    def fetch_items(self) -> list[ContentData]:
        """Fetch video transcripts."""
        video_ids = self._get_video_ids()
        logger.info(f"Processing {len(video_ids)} videos")

        items = []
        for video_id in video_ids:
            # Get transcript
            transcript = self._fetch_transcript(video_id)
            if not transcript:
                logger.info(f"Skipping {video_id} - no transcript available")
                continue

            # Get metadata
            metadata = self._fetch_video_metadata(video_id)

            items.append(
                ContentData(
                    external_id=video_id,
                    title=metadata.get("title", f"Video {video_id}"),
                    author=metadata.get("author_name"),
                    content_text=transcript,
                    content_html=None,
                    url=f"https://www.youtube.com/watch?v={video_id}",
                    published_at=None,  # oembed doesn't provide publish date
                )
            )

        logger.info(f"Successfully fetched {len(items)} video transcripts")
        return items

    def ingest(self) -> IngestResult:
        """Ingest YouTube video transcripts."""
        result = IngestResult(source_id=self.source_id)

        # Validate config
        errors = self.validate_config()
        if errors:
            result.errors = errors
            return result

        try:
            items = self.fetch_items()
            result.items_found = len(items)
        except Exception as e:
            logger.error(f"Failed to fetch YouTube: {e}")
            result.errors.append(f"Failed to fetch YouTube: {e}")
            return result

        # Store items in database
        with get_session() as session:
            for item in items:
                try:
                    existing = (
                        session.query(ContentItem)
                        .filter_by(source_id=self.source_id, external_id=item.external_id)
                        .first()
                    )

                    if existing:
                        result.items_skipped += 1
                        continue

                    content_item = ContentItem(
                        source_id=self.source_id,
                        external_id=item.external_id,
                        title=item.title,
                        author=item.author,
                        content_text=item.content_text,
                        content_html=item.content_html,
                        url=item.url,
                        published_at=item.published_at,
                    )
                    session.add(content_item)
                    result.items_new += 1

                except Exception as e:
                    logger.error(f"Failed to store item {item.external_id}: {e}")
                    result.items_failed += 1
                    result.errors.append(f"Failed to store {item.title}: {e}")

        logger.info(
            f"YouTube ingestion complete: {result.items_new} new, "
            f"{result.items_skipped} skipped, {result.items_failed} failed"
        )
        return result

File: weekly-intel-web/src/lib/services/ingestion/youtube.ts
import { db } from "@/lib/db";
import { contentItems } from "@/lib/db/schema";
import { eq, and } from "drizzle-orm";

interface IngestResult {
  sourceId: number;
  itemsFound: number;
  itemsNew: number;
  itemsSkipped: number;
  itemsFailed: number;
  errors: string[];
}

interface YouTubeConfig {
  channelId?: string;
  channelUrl?: string;
  playlistId?: string;
  playlistUrl?: string;
  videoUrls?: string[];
  videoIds?: string[];
  maxVideos?: number;
}

function extractVideoId(url: string): string | null {
  const patterns = [
    /(?:v=|\/v\/|youtu\.be\/)([a-zA-Z0-9_-]{11})/,
    /(?:embed\/)([a-zA-Z0-9_-]{11})/,
    /(?:shorts\/)([a-zA-Z0-9_-]{11})/,
  ];
  for (const pattern of patterns) {
    const match = url.match(pattern);
    if (match) return match[1];
  }
  return null;
}

function extractPlaylistId(url: string): string | null {
  try {
    const parsed = new URL(url);
    return parsed.searchParams.get("list");
  } catch {
    return null;
  }
}

async function fetchPlaylistVideos(playlistId: string): Promise<string[]> {
  try {
    const res = await fetch(
      `https://www.youtube.com/playlist?list=${playlistId}`
    );
    const html = await res.text();
    const matches = html.match(/"videoId":"([a-zA-Z0-9_-]{11})"/g) ?? [];
    const ids = matches.map(
      (m) => m.match(/"videoId":"([^"]+)"/)![1]
    );
    // Deduplicate preserving order
    return [...new Set(ids)].slice(0, 50);
  } catch {
    return [];
  }
}

async function fetchChannelVideos(channelId: string): Promise<string[]> {
  try {
    const res = await fetch(
      `https://www.youtube.com/feeds/videos.xml?channel_id=${channelId}`
    );
    const xml = await res.text();
    // Simple XML parsing for video IDs
    const matches =
      xml.match(/yt:videoId>([a-zA-Z0-9_-]{11})</g) ?? [];
    return matches.map((m) => m.replace("yt:videoId>", "").replace("<", ""));
  } catch {
    return [];
  }
}

async function fetchTranscript(videoId: string): Promise<string | null> {
  try {
    // Fetch the watch page to get caption tracks
    const res = await fetch(`https://www.youtube.com/watch?v=${videoId}`);
    const html = await res.text();

    // Extract caption URL from page data
    const captionMatch = html.match(
      /"captionTracks":\[.*?"baseUrl":"([^"]+)"/
    );
    if (!captionMatch) return null;

    const captionUrl = captionMatch[1].replace(/\\u0026/g, "&");
    const captionRes = await fetch(captionUrl);
    const captionXml = await captionRes.text();

    // Parse caption XML
    const texts = captionXml.match(/<text[^>]*>([\s\S]*?)<\/text>/g) ?? [];
    const transcript = texts
      .map((t) => {
        const content = t.replace(/<[^>]+>/g, "");
        return content
          .replace(/&amp;/g, "&")
          .replace(/&lt;/g, "<")
          .replace(/&gt;/g, ">")
          .replace(/&#39;/g, "'")
          .replace(/&quot;/g, '"');
      })
      .join(" ");

    return transcript || null;
  } catch {
    return null;
  }
}

async function fetchVideoMetadata(
  videoId: string
): Promise<{ title: string; author: string | null }> {
  try {
    const res = await fetch(
      `https://www.youtube.com/oembed?url=https://www.youtube.com/watch?v=${videoId}&format=json`
    );
    if (res.ok) {
      const data = await res.json();
      return {
        title: data.title ?? `Video ${videoId}`,
        author: data.author_name ?? null,
      };
    }
  } catch {
    // ignore
  }
  return { title: `Video ${videoId}`, author: null };
}

async function getVideoIds(config: YouTubeConfig): Promise<string[]> {
  const videoIds: string[] = [];

  if (config.videoIds) videoIds.push(...config.videoIds);

  if (config.videoUrls) {
    for (const url of config.videoUrls) {
      const vid = extractVideoId(url);
      if (vid) videoIds.push(vid);
    }
  }

  let playlistId = config.playlistId ?? null;
  if (!playlistId && config.playlistUrl) {
    playlistId = extractPlaylistId(config.playlistUrl);
  }
  if (playlistId) {
    const ids = await fetchPlaylistVideos(playlistId);
    videoIds.push(...ids);
  }

  const channelId = config.channelId ?? null;
  if (channelId) {
    const ids = await fetchChannelVideos(channelId);
    videoIds.push(...ids);
  }

  // Deduplicate
  const unique = [...new Set(videoIds)];
  const maxVideos = config.maxVideos ?? 20;
  return unique.slice(0, maxVideos);
}

export async function ingestYouTube(
  sourceId: number,
  sourceConfig: YouTubeConfig,
  force = false
): Promise<IngestResult> {
  const result: IngestResult = {
    sourceId,
    itemsFound: 0,
    itemsNew: 0,
    itemsSkipped: 0,
    itemsFailed: 0,
    errors: [],
  };

  const hasSource =
    sourceConfig.channelId ||
    sourceConfig.channelUrl ||
    sourceConfig.playlistId ||
    sourceConfig.playlistUrl ||
    sourceConfig.videoUrls?.length ||
    sourceConfig.videoIds?.length;

  if (!hasSource) {
    result.errors.push(
      "YouTube source requires one of: channelId, channelUrl, playlistId, playlistUrl, videoUrls, or videoIds"
    );
    return result;
  }

  let videoIds: string[];
  try {
    videoIds = await getVideoIds(sourceConfig);
  } catch (e) {
    result.errors.push(`Failed to get video IDs: ${e}`);
    return result;
  }

  result.itemsFound = videoIds.length;

  for (const videoId of videoIds) {
    try {
      // Check if exists
      const existing = await db
        .select({ id: contentItems.id })
        .from(contentItems)
        .where(
          and(
            eq(contentItems.sourceId, sourceId),
            eq(contentItems.externalId, videoId)
          )
        )
        .limit(1);

      if (existing.length > 0 && !force) {
        result.itemsSkipped++;
        continue;
      }

      const transcript = await fetchTranscript(videoId);
      if (!transcript) {
        result.itemsSkipped++;
        continue;
      }

      const metadata = await fetchVideoMetadata(videoId);

      if (existing.length > 0 && force) {
        await db
          .update(contentItems)
          .set({
            title: metadata.title,
            author: metadata.author,
            contentText: transcript,
            url: `https://www.youtube.com/watch?v=${videoId}`,
          })
          .where(eq(contentItems.id, existing[0].id));
        result.itemsSkipped++;
      } else {
        await db.insert(contentItems).values({
          sourceId,
          externalId: videoId,
          title: metadata.title,
          author: metadata.author,
          contentText: transcript,
          url: `https://www.youtube.com/watch?v=${videoId}`,
        });
        result.itemsNew++;
      }
    } catch (e) {
      result.itemsFailed++;
      result.errors.push(`Failed to process video ${videoId}: ${e}`);
    }
  }

  return result;
}
